openmdao/utils/reports_system.py

```diff
@@ -80,8 +115,69 @@ def register_report(name, func, desc, class_name, method, pre_or_post, filename=
         raise ValueError("The argument 'pre_or_post' can only have values of 'pre' or 'post', "
                          f"but {pre_or_post} was given")
 
-    _reports_registry[name] = _Report(func, desc, class_name, inst_id, method, pre_or_post,
-                                      filename)
+    _reports_registry[name] = report = _Report(name, filename, desc)
+
+    pre = func if pre_or_post == 'pre' else None
+    post = func if pre_or_post == 'post' else None
+    report.register_hook(fname=method, class_name=class_name, inst_id=inst_id, pre=pre, post=post,
+                         ncalls=1, **kwargs)
+
+
+def unregister_report(name):
+    """
+    Unregister a report with the reporting system.
+
+    Parameters
+    ----------
+    name : str
+        Name of report. Report names must be unique across all reports.
+    """
+    global _reports_registry
+    del _reports_registry[name]
+
+
+def register_report_hook(name, fname, class_name, inst_id=None, pre=None, post=None, filename=None,
+                         description='', **kwargs):
+    """
+    Register a hook with a specific report name in the reporting system.
+
+    By calling this multiple times, multiple hooks can be registered for the same report. This
+    is sometimes necessary to get the correct behavior when setup errors occur prior to report
+    creation.
+
+    Parameters
+    ----------
+    name : str
+        Name of report. Report names must be unique across all reports.
+    fname : str
+        The name of the function where the pre and/or post hook will be applied.
+    class_name : str
+        The name of the class owning the method where the hook will be applied.
+    inst_id : str or None
+        The name of the instance owning the method where the hook will be applied.
+    pre : function (None)
+        If not None, this hook will run before the named function runs.
+    post : function (None)
+        If not None, this hook will run after the named function runs.
```

Comment:
pedantic nit...  these arguments are a function _reference_ not a function name (so "after the referenced function"...)

Also,  should this function be added to the doc page for the Reports System?